be/src/http/action/stream_load.cpp

```diff
@@ -299,23 +307,40 @@ void StreamLoadAction::on_chunk_data(HttpRequest* req) {
     auto evbuf = evhttp_request_get_input_buffer(ev_req);
 
     int64_t start_read_data_time = MonotonicNanos();
-    while (evbuffer_get_length(evbuf) > 0) {
-        ByteBufferPtr bb = ByteBuffer::allocate(4096);
+
+    size_t len = 0;
+    while ((len = evbuffer_get_length(evbuf)) > 0) {
+        if (ctx->buf->remaining() < len) {
+            if (ctx->format == TFileFormatType::FORMAT_JSON) {
+                // For json format, we need build a complete json before we push the buffer to the pipe.
+                // buffer capacity is not enough, so we try to expand the buffer.
+                ByteBufferPtr buf = ByteBuffer::allocate(BitUtil::RoundUpToPowerOfTwo(ctx->buf->pos + len));
+                buf->put_bytes(ctx->buf->ptr, ctx->buf->pos);
+                std::swap(buf, ctx->buf);
+
+            } else {
+                // For non-json format, we could push buffer to the body_sink in streaming mode.
+                // buffer capacity is not enough, so we push the buffer to the pipe and allocate new one.
+                ctx->buf->flip();
+                auto st = ctx->body_sink->append(std::move(ctx->buf));
+                if (!st.ok()) {
+                    LOG(WARNING) << "append body content failed. errmsg=" << st.get_error_msg() << ctx->brief();
+                    ctx->status = st;
+                    return;
+                }
+
+                ctx->buf = ByteBuffer::allocate(ctx->buf_default_sz);
```

Comment:
```suggestion
                ctx->buf = ByteBuffer::allocate(std::max(len, ctx->buf_default_sz));
```

Reply:
`len` denotes the length of current returned data, but the `ctx->buf` is prepared for next read. What's your consideration here?

Reply:
`ctx->buf` is used to read from `evbuf` through `evbuffer_remove`, according to the following code at line 341, if the capacity of `ctx->buf` is less than `len` the `evbuf` cannot be read fully at once, right?

Reply:
Got that. Updated.